webapps-backup/src/main/java/io/camunda/webapps/backup/repository/opensearch/OpensearchBackupRepository.java

```diff
@@ -0,0 +1,475 @@
+      throw new InvalidRequestException(reason);
+    }
+  }
+
+  @Override
+  public GetBackupStateResponseDto getBackupState(
+      final String repositoryName, final Long backupId) {
+    final List<OpenSearchSnapshotInfo> snapshots = findSnapshots(repositoryName, backupId);
+    return toGetBackupStateResponseDto(backupId, snapshots);
+  }
+
+  @Override
+  public List<GetBackupStateResponseDto> getBackups(final String repositoryName) {
+    final var requestBuilder =
+        getSnapshotRequestBuilder(repositoryName, snapshotNameProvider.snapshotNamePrefix() + "*")
+            .build();
+    final OpenSearchGetSnapshotResponse response;
+    try {
+      response =
+          OpenSearchGetSnapshotResponse.fromResponse(
+              openSearchClient.snapshot().get(requestBuilder));
+      final List<OpenSearchSnapshotInfo> snapshots =
+          response.snapshots().stream()
+              .sorted(Comparator.comparing(OpenSearchSnapshotInfo::getStartTimeInMillis).reversed())
+              .toList();
+
+      final LinkedHashMap<Long, List<OpenSearchSnapshotInfo>> groupedSnapshotInfos =
+          snapshots.stream()
+              .collect(
+                  groupingBy(
+                      si -> {
+                        final Metadata metadata =
+                            Metadata.fromMetadata(
+                                si.getMetadata(), openSearchClient._transport().jsonpMapper());
+                        Long backupId = metadata.backupId();
+                        // backward compatibility with v. 8.1
+                        if (backupId == null) {
+                          backupId = snapshotNameProvider.extractBackupId(si.getSnapshot());
+                        }
+                        return backupId;
+                      },
+                      LinkedHashMap::new,
+                      toList()));
+
+      return groupedSnapshotInfos.entrySet().stream()
+          .map(entry -> toGetBackupStateResponseDto(entry.getKey(), entry.getValue()))
+          .toList();
+
+    } catch (final Exception e) {
+      if (isRepositoryMissingException(e)) {
+        final String reason = noRepositoryErrorMessage(repositoryName);
+        throw new GenericBackupException(reason);
+      }
+      if (isSnapshotMissingException(e)) {
+        // no snapshots exist
+        return new ArrayList<>();
+      }
+      final String reason =
+          format("Exception occurred when searching for backups: %s", e.getMessage());
+      throw new GenericBackupException(reason, e);
+    }
+  }
+
+  @Override
+  public void executeSnapshotting(
+      final BackupService.SnapshotRequest snapshotRequest,
+      final Runnable onSuccess,
+      final Runnable onFailure) {
+    final Long backupId = backupId(snapshotRequest);
+    final Map<String, JsonData> metadataJson =
+        snapshotRequest.metadata().asJson(openSearchClient._transport().jsonpMapper());
```

Comment:
Deserializing Metadata with the custom logic, not using `objectMapper` injected from outside the context, see above comment